plugins/UsersManager/API.php

```diff
@@ -107,7 +107,16 @@ public function getUserPreference($userLogin, $preferenceName)
             return $optionValue;
         }
 
-        return $this->getDefaultUserPreference($preferenceName, $userLogin);
+        $defaultPreference = $this->getDefaultUserPreference($preferenceName, $userLogin);
+
+        // A `get*` method should not set anything. But in this case we do it to improve performance as it prevents
+        // us from loading all siteIds (which can be 50k or more) the next time this preference is requested.
+        // this method can be called quite often when generating links etc (to get defaultWebsiteId).
+        if ($defaultPreference !== false && $preferenceName === self::PREFERENCE_DEFAULT_REPORT) {
+            $this->setUserPreference($userLogin, $preferenceName, $defaultPreference);
+        }
```

Comment:
The current approach of hard-coding a set in the get... method has the side effect of adding technical debt to a part of Piwik that is already tightly coupled w/ core. And the code should never be run more than once per user (unless the option table was deleted or something), yet its in a method that will be executed far more often.

I think for this particular change, we should find another way to apply it, though I'm not sure exactly what to do. If there's no way to do this w/o doing it on every request, perhaps we can use an event handler in UsersManager? eg, Platform.initialized or something. I'm not sure how early this has to be executed.

Re time: if we can agree on a way to change it, I wouldn't mind implementing it myself.